// nexmo/lib/ShortCode.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _Utils = require("./Utils");

var _Utils2 = _interopRequireDefault(_Utils);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var querystring = require("querystring");

/**
 * Provides access to the `ShortCode` endpoint.
 */

var ShortCode = function () {
  _createClass(ShortCode, null, [{
    key: "PATH",
    get: function get() {
      return "/sc/us/${type}/json";
    }
  }, {
    key: "ERROR_MESSAGES",
    get: function get() {
      return {
        to: "Invalid to address",
        msgParams: "Invalid shortcode message parameters"
      };
    }

    /**
     * @param {Credentials} credentials
     *    credentials to be used when interacting with the API.
     * @param {Object} options
     *    Additional ShortCode options.
     */

  }]);

  function ShortCode(credentials) {
    var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

    _classCallCheck(this, ShortCode);

    this.creds = credentials;
    this.options = options;
  }

  _createClass(ShortCode, [{
    key: "_sendRequest",
    value: function _sendRequest(endpoint, method, callback) {
      endpoint.path = endpoint.path + (endpoint.path.indexOf("?") > 0 ? "&" : "?") + querystring.stringify({
        api_key: this.creds.apiKey,
        api_secret: this.creds.apiSecret
      });
      this.options.httpClient.request(endpoint, method, callback);
    }
  }, {
    key: "_sendViaShortcode",
    value: function _sendViaShortcode(type, recipient, messageParams, opts, callback) {
      if (!recipient) {
        _Utils2.default.sendError(callback, new Error(ShortCode.ERROR_MESSAGES.to));
      }
      if (!messageParams || !Object.keys(messageParams)) {
        _Utils2.default.sendError(callback, new Error(ShortCode.ERROR_MESSAGES.msgParams));
      }
      opts = opts || {};
      var path = ShortCode.PATH.replace("${type}", type);
      Object.keys(messageParams).forEach(function (key) {
        opts[key] = messageParams[key];
      });
      opts.to = recipient;
      path += "?" + querystring.stringify(opts);
      this.options.logger.info("sending message from shortcode " + type + " to " + recipient + " with parameters " + JSON.stringify(messageParams));
      this._sendRequest({
        host: this.options.restHost || "rest.nexmo.com",
        path: path
      }, "POST", function (err, apiResponse) {
        if (!err && apiResponse.status && apiResponse.messages[0].status > 0) {
          _Utils2.default.sendError(callback, new Error(apiResponse.messages[0]["error-text"]), apiResponse);
        } else {
          if (callback) callback(err, apiResponse);
        }
      });
    }
  }, {
    key: "shortcodeAlert",
    value: function shortcodeAlert(recipient, messageParams, opts, callback) {
      this._sendViaShortcode("alert", recipient, messageParams, opts, callback);
    }
  }, {
    key: "shortcode2FA",
    value: function shortcode2FA(recipient, messageParams, opts, callback) {
      this._sendViaShortcode("2fa", recipient, messageParams, opts, callback);
    }
  }, {
    key: "shortcodeMarketing",
    value: function shortcodeMarketing(recipient, messageParams, opts, callback) {
      this._sendViaShortcode("marketing", recipient, messageParams, opts, callback);
    }
  }]);

  return ShortCode;
}();

exports.default = ShortCode;
module.exports = exports["default"];
